File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    void solve(int idx,vector<int>& candidates,vector<int>& combination,int target,vector<vector<int>>&ans,int sumSoFar){
        int n = candidates.size();
        
        if(idx==n){
            if(sumSoFar==target) ans.push_back(combination);
        }
        else if(target>=sumSoFar){
            solve(idx+1,candidates,combination,target,ans,sumSoFar);
            combination.push_back(candidates[idx]);
            solve(idx,candidates,combination,target,ans,sumSoFar+candidates[idx]);
            combination.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> combination;
        solve(0,candidates,combination,target,ans,0);
        return ans;
    }
};

File: 39-combination-sum/39-combination-sum.py
class Solution:
    def solve(self,nums,target,so_far):
        if(target==0): self.ans.append(so_far)
        if(target>0):
            if(len(nums)>1):
                self.solve(nums[1:],target,so_far)
            self.solve(nums,target-nums[0],so_far+[nums[0]])
        
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.ans=[]
        self.solve(candidates,target,[])
        return self.ans